### ISSUE TITLE
Adding gcc-8 case on Travis CI.

### ISSUE THREAD

**Issue by junaruga** [BenLangmead/bowtie2#207]
Thank you for maintaining bowtie2.
I would like to propose to add gcc-8 test case on Travis CI.

## Background

Right now this repository has gcc test case on Travis CI.
The gcc version is 4.8 as the default setting. In other words, the latest version of gcc is 8.
Adding the latest version of gcc on the CI is helpful when people will use bowtie2 on latest platform.
In my opinion running both old version 4.8 and 8 on the CI is meaningful.

## Possible implementations

I prepared 2 patterns on my repository.
I want to ask you.

### 2. Which pattern is better for you?

2-a: The way to use Travis yaml syntax as much as possible.

https://github.com/junaruga/bowtie2/commit/a987ab9c2c6
https://travis-ci.org/junaruga/bowtie2/builds/427367041

2-b: The way to use shell commands as much as possible.

https://github.com/junaruga/bowtie2/commit/199d2f7
https://travis-ci.org/junaruga/bowtie2/builds/427097457

There is an issue for gcc-8 case.
gcc-8 total running time is 46 mins while gcc (4.8)'s total running time is 22 or 23 mins.
Maybe we need to fix the issue.

## References

You might be able to refer below documents.

Travis documents
https://docs.travis-ci.com/user/languages/c/#gcc-on-linux
https://docs.travis-ci.com/user/installing-dependencies/#adding-apt-sources-1

Case: Ruby
https://github.com/ruby/ruby/blob/trunk/.travis.yml
https://travis-ci.org/ruby/ruby

Case: Trinity
https://travis-ci.org/trinityrnaseq/trinityrnaseq
https://github.com/trinityrnaseq/trinityrnaseq/blob/master/.travis.yml

Thank you.

**Comment by ch4rr0** [comment]
I've added support for gcc-8 to the [travis config](https://github.com/BenLangmead/bowtie2/blob/master/.travis.yml). If you have any feedback please let me know.

**Comment by junaruga** [comment]
Thank you for the work!

It seems that the "travis config" on master branch does not have gcc-8 case.
https://github.com/BenLangmead/bowtie2/blob/master/.travis.yml

You meant the travis config on bug_fixes?
https://github.com/BenLangmead/bowtie2/blob/bug_fixes/.travis.yml

**Comment by ch4rr0** [comment]
Yes, just bug_fixes for now. We will merge the changes to master during the release process

**Comment by junaruga** [comment]
Sure. I am happy to give you some feedback.

**Comment by junaruga** [comment]
Now there are some feedback, seeing the bug_fixes branch's `.travis.yml`, the [bug_fixes branch's Travis result](https://travis-ci.org/BenLangmead/bowtie2/builds/632368221).

* `before-install:` is wrong. `before_install:` (underscore) works. Right now the `before-install:` part is actually not executed.
* You do not need to use `MATRIX_EVAL` to use `gcc-4.9` or `gcc-8`, when you set `compiler: gcc-4.9` or `compiler: gcc-8`. When I tried the way without using `MATRIX_EVAL` on my forked repository's Travis CI, it worked. This way is reliable, as it has been used in [ruby project's .travis.yml](https://github.com/ruby/ruby/blob/master/.travis.yml#L64).
* You are running aarch64 case in old way: using container. As you may know, you need to rebase bug_fixes branch on the latest master branch. To run aarch64 case with new way `arch: arm64`, you need to merge the commits https://github.com/BenLangmead/bowtie2/commit/c6a6648fcd05bff0e5ca64a305bb1c488beacbcf  and https://github.com/BenLangmead/bowtie2/commit/2df34deb6cc71333e32f860f7c505d12f9330eb2 to bug_fixes branch. To run the new aarch64 case, please check below my `.travis.yml` in Enhancements section.

### Aligning YAML indent format.

In the `.travis.yml` there are 2 patterns for the YAML indent. Both are valid.

Pattern A

```
foo: bar
  - a
  - b
```

Pattern B

```
foo: bar
- a
- b
````

You modified with the pattern A. And I like the pattern A.
Following pattern B parts can be modified.

```
branches:
  only:
  - master
  - cleaner_parsing
  - scaling-old-parsing
  - batch_parsing
  - batch_parsing_output
  - blocked_input
  - bug_fixes
```

```
          packages:
          - tbb 
```

### Enhancements

To explain it, I prepared [.traivs.yml](https://github.com/junaruga/bowtie2/blob/feature/update-travis/.travis.yml). Seeing the file, please read following text. `branches:` and `notifications:` are commented out in the file, because I wanted to run Travis on my forked repository.

The file is [one commit](https://github.com/junaruga/bowtie2/commits/feature/update-travis
) on the latest master branch of my forked repository, and based on your bug_fixes branche's `.travis.yml`. Here is the [Travis result](https://travis-ci.org/junaruga/bowtie2/builds/632746737).

* It is safe to set `dist: xenial` explicitly in `.travis.yml`. Because while right now the Travis's `dist`'s default value is `xenial`, it is possible to change the value in the future.

* I removed following `CXXFLAGS` setting in aarch64 case.

  ```
  CXXFLAGS="-Wno-deprecated-declarations -Wno-misleading-indentation -Wno-narrowing -Wno-unused-function -Wno-unused-result"
  ```

  Because right now there is only one warning without the variable in aarch64 case, and it's not difficult to fix it. And the a few warnings do not disturb seeing the Travis log.

  ```
  alphabet.cpp:224:1: warning: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ inside { } is ill-formed in C++11 [-Wnarrowing]
   };
   ^
  ```

* `make -s -j $NPROC` to save the running time. `$NPROC` is the number of processors (CPU cores). This make command saves the running time. See the Travis CI result. I created `scripts/travis_wait.sh` script that is previously written directly in `.travis.yml`. Now it is used in the 2 parts of the `.travis.yml`.

* I moved `before_install:` to above `script`. Because `before_install:`, `install:`, `before_script:` and `script:` syntax are same kind of one. And it is executed by the order.

* I added `- uname -m` in `before_install` and `- echo CC=${CC} CXX=${CXX} NPROC=${NPROC}` in `before_script` is for convenience.

If you like, it is possible for me to sent the pull-request about the `.travis.yml`.

**Comment by junaruga** [comment]
I prepared another [.travis.yml](https://github.com/junaruga/bowtie2/blob/feature/update-travis-python3/.travis.yml) on my forked repository 's [update-travis-python3 branch](https://github.com/junaruga/bowtie2/tree/feature/update-travis-python3) as one commit.

Here is [the Travis CI result](https://travis-ci.org/junaruga/bowtie2/builds/633247707).

I think this `.travis.yml` is better than my above one.

There are 4 jobs.
For 1st and 2nd job, `language: python` mode is used to specify `python` to the version.
For 3rd and 4th job, `language: cpp` is used.
The point is we can not basically use multiple `language: foo` for a job except a few limited case (`language: ruby ` + `jdk: foo` (`jdk:` is the syntax of `language: java` works)

1. x86_64, Linux, gcc-4.8 with python 3.4 (python 3 old version)
2. x86_64, Linux, gcc-8 with python 3.8 (python 3 latest stable new version)
3. x86_64, OSX clang with default python (= python 2.7)
4. ARM64, Linux, default gcc version (= gcc 5.4) with default python (= python 2.7)

Other changes:
* I removed common logic's `make` command's `-s` option again from above my [.travis.yml](https://github.com/junaruga/bowtie2/blob/feature/update-travis/.travis.yml) on update-travis branch.
* I added the logic to print C/C++ compiler and python version.

```
before_script:
  - uname -m
  - python --version
  - echo CC=${CC} CXX=${CXX} NPROC=${NPROC}
  - $CC --version
  - $CXX --version
script: scripts/travis_wait.sh make -j $NPROC allall && make simple-test
```

**Comment by ch4rr0** [comment]
Thank you for the feedback. These are some good changes and I will incorporate them.

**Comment by ch4rr0** [comment]
We've worked together on incorporating these changes and they should be available in the  v2.4.0 and later. Feel free to reopen this issue if there is anything you feel was not addressed.
